working/0546-patterns/patterns-feature-specification.md

```diff
@@ -2159,17 +2209,233 @@ To match a pattern `p` against a value `v`:
+matching in Dart for the kinds of objects they already use, including mutable
+lists and maps and instances of user-defined classes whose getters can't be
+proven to be pure and side-effect free. At the same time, we also want to ensure
+that exhaustiveness checking is correct and sound.
+
+To balance those, pattern matching operates on an *immutable snapshot of the
+properties of the matched value that are seen by the patterns*. The way this
+works is that whenever a member is invoked on the matched value or an object
+returned by some previous destructuring, the result is cached. Whenever the same
+member is invoked by a later pattern (either a subsequent subpattern, or a
+pattern in a later case), we don't invoke the member again and instead use the
+previously returned value. This way, all subpatterns and cases see the exact
+same portions of the object and from the perspective of the surrounding switch
+statement or other construct, the object appears to be immutable.
+
+For example, consider:
+
+```dart
+main() {
+  var list = [1, 2];
+  switch (list) {
+    case [1, _] & [_, < 4]: print('first');
+    case [int(isEven: true), var a]: print('second $a');
+  }
+}
+```
+
+As written, there appear to be multiple redundant method calls on `list` and the
+elements extracted from it. But the actual execution semantics are roughly like:
+
+```dart
+main() {
+  var list = [1, 2];
+
+  late final $match = list;
+  late final $match_length = $match.length;
+  late final $match_length_eq2 = $match_length == 2;
+  late final $match_0 = $match[0];
+  late final $match_1 = $match[1];
+  late final $match_0_eq1 = $match_0 == 1;
+  late final $match_1_lt4 = $match_1 < 4;
+  late final $match_0_isEven = $match_1.isEven;
+  late final $match_0_isEven_eqtrue = $match_0_isEven == true;
+
+  if ($match_length_eq2 &&
+      $match_0_eq1 &&
+      $match_length_eq2 &&
+      $match_1_lt4) {
+    print('first');
+  } else if ($match_length_eq2 &&
+      $match_0_isEven_eqtrue) {
+    var a = $match_1;
+    print('second $a');
+  }
+}
+```
+
+Note that every method call is encapsulated in a `late` variable ensuring that
+it only gets invoked once even when used by multiple patterns.
+
+It works like this:
+
+1.  At compile time, after type checking has completed, we associate an
+    *invocation key* with every member call or record field access potentially
+    made by each pattern.
+
+2.  At runtime, whenever the runtime semantics say to call a member or access a
+    record field, if a previous call or access with that same invocation key has
+    already been evaluated, we reuse the result.
+
+3.  Otherwise, we invoke the member or field access now and associate the result
+    with that invocation key for future calls.
+
+Let an *invocation key* comprise:
+
+*   A possibly absent parent invocation key.
+*   A possibly absent extension type. If the invocation represents an extension
```

Comment:
"Extension type" also confuses me a little, because extensions are not types.
I assume it covers an extension declaration and the inferred type parameters to that extension. Could it be called "extension instantiation"?

Reply:
Reworded.